src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+            // moving on -- clear frontier
+            self.frontier.clear();
+        }
+
+        self.prev_neighbor = Some(self.cur_neighbor.clone());
+        self.cur_neighbor = next_neighbor;
+        self.walk_outbound = next_neighbor_outbound;
+        self.next_neighbor = None;
+
+        self.clear_connections(local_peer);
+        self.new_frontier.clear();
+
+        let result = self.result.clone();
+
+        self.walk_end_time = get_epoch_time_secs();
+
+        // leave self.frontier and self.result alone until the next walk.
+        // (makes it so that at the end of the walk, we can query the result and frontier, which
+        // get built up over successive passes of the state-machine)
+        result
+    }
+
+    /// Clear the walk's connection state
+    fn clear_connections(&mut self, _local_peer: &LocalPeer) -> () {
+        test_debug!("{:?}: Walk clear connections", _local_peer);
+        self.events.clear();
+        self.connecting.clear();
+        self.pending_neighbor_addrs = None;
+
+        self.ongoing_request = None;
+        self.ongoing_batch_request = None;
+
+        self.resolved_handshake_neighbors.clear();
+        self.handshake_neighbor_addrs.clear();
+
+        self.resolved_getneighbors_neighbors.clear();
+
+        self.neighbor_replacements.clear();
+        self.replaced_neighbors.clear();
+
+        self.network_pingbacks.clear();
+    }
+
+    /// Update the state of the walk.
+    /// If the code spent too much time in one state, then the walk will fail with StepTimeout
+    fn set_state(
+        &mut self,
+        _local_peer: &LocalPeer,
+        new_state: NeighborWalkState,
+    ) -> Result<(), net_error> {
+        if self.walk_state_time + self.walk_state_timeout < get_epoch_time_secs() {
+            return Err(net_error::StepTimeout);
+        }
+
+        test_debug!(
+            "{:?}: Advance walk state: {:?} --> {:?} (after {} seconds)",
+            _local_peer,
+            &self.state,
+            &new_state,
+            get_epoch_time_secs().saturating_sub(self.walk_state_time)
+        );
+        self.state = new_state;
+        self.walk_state_time = get_epoch_time_secs();
+        Ok(())
+    }
+
+    /// Begin handshaking with our current neighbor.
+    /// On success, return Ok(true) and transition to HandshakeFinish
+    /// On failure, return an Err(...)
+    /// If we're not yet connected, return Ok(false).  The caller should try again.
+    pub fn handshake_begin(&mut self, network: &mut PeerNetwork) -> Result<bool, net_error> {
+        if self.ongoing_request.is_some() {
+            // in progress already
+            return Ok(true);
+        }
+
+        // if cur_neighbor is _us_, then grab a different neighbor and try again
+        if Hash160::from_node_public_key(&self.cur_neighbor.public_key)
+            == Hash160::from_node_public_key(&Secp256k1PublicKey::from_private(
+                &network.get_local_peer().private_key,
+            ))
```

Comment:
The function `Hash160::from_node_public_key()` evaluates to the hash160 of the compressed public key, even if the argument is not compressed.  It's defensive programming.

Reply:
That's a pretty surprising usage of these methods -- that should either have a comment like:

```rust
// compare the Hash160's because `Hash160::from_node_public_key` forces comparison of the compressed public // keys
```

So that a future maintainer knows why this is being done, or it should be replaced with an explicit force-to-compressed-key method.

Reply:
I'll add a comment.